### PR TITLE
caddyhttp: New placeholder for a PEM encoded value of client certificate

### PR DESCRIPTION
**Pull request by gdhameeja**
For #3585

### CODE REVIEW

File: modules/caddyhttp/replacer.go [hunk 1]
```diff
@@ -343,6 +344,9 @@ func getReqTLSReplacement(req *http.Request, key string) (interface{}, bool) {
 			return cert.SerialNumber, true
 		case "client.subject":
 			return cert.Subject, true
+		case "client.certificate_pem":
+			block := pem.Block{Type: "message", Headers: map[string]string{}, Bytes: cert.Raw}
```
Comment (mholt) [inline modules/caddyhttp/replacer.go:348]:
```suggestion
			block := pem.Block{Type: "CERTIFICATE", Bytes: cert.Raw}
```

File: modules/caddyhttp/replacer_test.go [hunk 2]
```diff
@@ -171,6 +171,10 @@ eqp31wM9il1n+guTNyxJd+FzVAH+hCZE5K+tCgVDdVFUlDEHHbS/wqb2PSIoouLV
 			input:  "{http.request.tls.client.san.ips.0}",
 			expect: "127.0.0.1",
 		},
+		{
+			input:  "{http.request.tls.client.certificate_pem}",
+			expect: "<empty>",
```
Comment (mholt) [inline modules/caddyhttp/replacer_test.go:176]: This doesn't seem like the right expected result... it should be the same as the `clientCert` var defined above.

**Comment by gdhameeja** [comment]
> Thanks for working on this! It needs some more work before it is ready to be merged.

If you are referring to "whether the client certificate was verified or not (in case of client_auth mode being request)", could I work on it? With some guidance ofcourse. Thanks

File: modules/caddyhttp/replacer_test.go (hunk 2 above)
Comment (gdhameeja) [reply modules/caddyhttp/replacer_test.go]: Fixed. Although the PEM encoded value comes with a newline attached due to which tests were failing.

**Comment by gdhameeja** [comment]
> Can you verify that tests fail if the expect is anything different?
```
$ go test .
--- FAIL: TestHTTPVarReplacement (0.00s)
    replacer_test.go:181: Test 24: Expected placeholder {http.request.tls.client.certificate_pem} to be 'gauravdhameeja
        ' but got '-----BEGIN CERTIFICATE-----
        MIIB9jCCAV+gAwIBAgIBAjANBgkqhkiG9w0BAQsFADAYMRYwFAYDVQQDDA1DYWRk
        eSBUZXN0IENBMB4XDTE4MDcyNDIxMzUwNVoXDTI4MDcyMTIxMzUwNVowHTEbMBkG
        A1UEAwwSY2xpZW50LmxvY2FsZG9tYWluMIGfMA0GCSqGSIb3DQEBAQUAA4GNADCB
        iQKBgQDFDEpzF0ew68teT3xDzcUxVFaTII+jXH1ftHXxxP4BEYBU4q90qzeKFneF
        z83I0nC0WAQ45ZwHfhLMYHFzHPdxr6+jkvKPASf0J2v2HDJuTM1bHBbik5Ls5eq+
        fVZDP8o/VHKSBKxNs8Goc2NTsr5b07QTIpkRStQK+RJALk4x9QIDAQABo0swSTAJ
        BgNVHRMEAjAAMAsGA1UdDwQEAwIHgDAaBgNVHREEEzARgglsb2NhbGhvc3SHBH8A
        AAEwEwYDVR0lBAwwCgYIKwYBBQUHAwIwDQYJKoZIhvcNAQELBQADgYEANSjz2Sk+
        eqp31wM9il1n+guTNyxJd+FzVAH+hCZE5K+tCgVDdVFUlDEHHbS/wqb2PSIoouLV
        3Q9fgDkiUod+uIK0IynzIKvw+Cjg+3nx6NQ0IM0zo8c7v398RzB4apbXKZyeeqUH
        9fNwfEi+OoXR6s+upSKobCmLGLGi9Na5s5g=
        -----END CERTIFICATE-----
        '
FAIL
FAIL	github.com/caddyserver/caddy/v2/modules/caddyhttp	0.031s
FAIL
```
yeah, it fails if expected is anything else.

**Comment by gdhameeja** [comment]
Signed the CLA as well:
image

**Comment by francislavoie** [comment]
I think the email you used to sign the CLA doesn't match the email in your commits, hence the CLA assistant complaining. Please either add the email address that you made your commits with to your github account, or rebase/rewrite your commits with the correct email address after editing your git config.

**Comment by gdhameeja** [comment]
@francislavoie Thanks for the suggestion. Learned something new today :)

**Comment by francislavoie** [comment]
With "Closes", it will automatically close that issue when this is merged (hover over the word closes in your top comment). If something is only a partial fix, then don't use the word "closes", but rather something like "For #<issue>" or "Partly completes #<issue>"

**Comment by gdhameeja** [comment]
@francislavoie I didn't know that, I will edit my comment as per your suggestion.

**Review by mholt — APPROVED** [review]
Awesome, thanks! Sorry it took me a while to circle back on this.

**Pull request merged**